### PR TITLE
🗂 Add site option to include folder structure in url paths

### PR DESCRIPTION
**Pull request by fwkoch**
This PR adds an option to support nested folder structure in URLs. See https://github.com/jupyter-book/mystmd/issues/670

This uses an option at the `site` level of your `myst.yml` - `site.options.url_folders: true`. Given this option, slugs are computed to contain dot-delimited folder structure (as opposed to the previous behavior where slugs were computed entirely based on filename). Consuming these new slugs with folders requires an updated myst theme, see https://github.com/jupyter-book/myst-theme/pull/489 (without the updated theme, the dots in the slugs will just show up in your MyST site; with the updated theme, your site URLs will have slashes instead of dots)

### CODE REVIEW

File: packages/myst-cli/src/project/load.ts [hunk 1]
```diff
@@ -57,9 +57,13 @@ export async function loadProjectFromDisk(
   let newProject: Omit<LocalProject, 'bibliography'> | undefined;
   let { index, writeTOC } = opts || {};
   let legacyToc = false;
+  const siteConfig = selectors.selectLocalSiteConfig(state, path);
+  const urlFolders = !!siteConfig?.options?.url_folders;
```
Comment (fwkoch) [inline packages/myst-cli/src/project/load.ts:61]: I'm happy to take suggestions about a better name for this option...

Comment (agoose77) [reply packages/myst-cli/src/project/load.ts:61]: @fwkoch do we have a policy about `snake_case` vs `pascalCase` for frontmatter/config item names?

**Comment by stevejpurves** [comment]
@fwkoch seems like there will be no impact on the `location` field in the output mdast and in `source` on notebook outputs?

**Comment by fwkoch** [comment]
@stevejpurves - yeah, `location` and `source` should be unchanged - these are still the relative file paths.

File: packages/myst-cli/src/process/mdast.ts [hunk 2]
```diff
@@ -262,6 +262,7 @@ export async function transformMdast(
     url = `/${useSlug ? pageSlug : ''}`;
     dataUrl = `/${pageSlug}.json`;
   }
+  url = url?.replace('.', '/');
```
Comment (rowanc1) [inline packages/myst-cli/src/process/mdast.ts:265]: do we need the `.index` and replaceAll?

File: packages/myst-cli/src/process/site.ts [hunk 3]
```diff
@@ -172,14 +172,14 @@ export async function writeMystSearchJson(session: ISession, pages: LocalProject
...
-      if (!mdast || !frontmatter) {
+      if (!mdast || !frontmatter || !slug) {
         return [];
       }
       const title = frontmatter.title ?? '';
 
       // Group by section (simple running accumulator)
       const sections = toSectionedParts(mdast);
-      const pageURL = slug && DEFAULT_INDEX_FILENAMES.includes(slug) ? '/' : `/${slug}`;
+      const pageURL = DEFAULT_INDEX_FILENAMES.includes(slug) ? '/' : `/${slug.replace('.', '/')}`;
```
Comment (rowanc1) [inline packages/myst-cli/src/process/site.ts:182]: replaceAll

File: packages/myst-cli/src/project/load.ts (hunk 1 above)
Comment (rowanc1) [reply packages/myst-cli/src/project/load.ts:61]: lets do `site.folders` and get around it!

File: packages/myst-cli/src/process/mdast.ts (hunk 2 above)
Comment (rowanc1) [reply packages/myst-cli/src/process/mdast.ts:265]:
```suggestion
  url = url?.replace(/\.index$/, '').replaceAll('.', '/');
```

File: packages/myst-cli/src/process/site.ts (hunk 3 above)
Comment (rowanc1) [reply packages/myst-cli/src/process/site.ts:182]:
```suggestion
      const pageURL = DEFAULT_INDEX_FILENAMES.includes(slug) ? '/' : `/${slug.replace(/\.index$/, '').replaceAll('.', '/')}`;
```

**Review by rowanc1 — APPROVED** [review]

**Pull request merged**